Compiler-Internals.md

```diff
@@ -18,19 +19,27 @@ The checker then waits until a particular symbol is requested to calculate type
 Each phase of the compiler (parser, binder, etc -- see below for details) treats data structures from the previous phases as immutable.
 In addition, data structures created within each phase are not usually modified after their creation.
 This requires a look-aside table in some cases.
-For example, because the binder only looks at one file at a time, the checker needs a merged-symbols table to track merged declarations. 
+For example, because the binder only looks at one file at a time, the checker needs a merged-symbols table to track merged declarations.
 It checks whether a symbol has an entry in the merged-symbols table each time before it uses a symbol.
 
 # Parser
 
+It's a recursive descent parser. It's pretty resilient, so if you search for functions matching the thing you want to change, you can probably get away with just adding the code to parse.
+There aren't any surprises in the general implementation style here.
+
 ## Incremental parsing
 
 ## ECMAScript parsing contexts
 
 ## ECMAScript automatic semicolon insertion
 
+## JSDoc parsing
+
 # Checker
 
+The checker is almost 20,000 lines long, and does almost everything that's not syntactic.
+Surprisingly, a checker gets created every time the language service requests information because it tries to present an immutable interface.
```

Comment:
the idea here is lifetimes and scopes of the different constructs. syntactic constructs have the same life time as the text they were created to represent. semantic constructs however are a function of a program, and any change to the program, be it adding or removing a file, or changing the text of one of the files, will invalidate that state, and it will need to be recreated.

One way of doing this is using mutable program objects, and then updating the state on every change. that proved to be a huge feet, and almost impossible to get right. so we opted into a simpler system, where the state is immutable, and on every change you create it again. 

We are willing to pay the cost of recreation; but in practice that does not end up being much, since the  whole system is realize everything lazily. so what you pay for is just the initialization of some basic types, and building the global symbol table.